### ISSUE TITLE
Adding Functional Group substitution to periodic crystals

### ISSUE THREAD

**Issue by computron** [materialsproject/pymatgen#1072]
## System

* Pymatgen version: master

## Summary

* Currently, one can only add a functional group to a site for a Molecule object (substitute() method)
* Sometimes, you want to do functional group replacement for molecular crystals (i.e. Structure objects)
* From what I can tell, this is easy to do. The substitute method essentially works out of the box, you just have to change:

```py
del self[index]
for site in func_grp[1:]:
    self._sites.append(site)
```

to

```py
del self[index]
for site in func_grp[1:]:
    s_new = PeriodicSite(site.species_and_occu, site.coords, self.lattice, coords_are_cartesian=True)
    self._sites.append(s_new)
```

Thoughts on:
* do you want to add this?
* what's the best way to do it to avoid code repetition, e.g. putting it in SiteCollection, etc.

**Comment by shyuep** [comment]
You can add this. If you can think of a general enough way to add it to SiteCollection, that would be good. But I think the Site vs PeriodicSite might be tricky.

The other thing we may want in future for both Mol and Structure is to be able to replace a functional group (e.g., -CH3) with a functional group (-NH2). Conceptually it isn't hard. There is already a method for bond breaking in Molecule. But doing it in a way that is robust is not easy.

**Comment by mkhorton** [comment]
I've been thinking of ways to do this also -- using the bond graph should make this relatively tractable, and this works with periodicity, might be overkill though. Will have a think about how to do this gracefully (i.e. integrating a common solution into SiteCollection). @espottesmith do you have any thoughts on this?

**Comment by espottesmith** [comment]
Warning: since I've been working pretty much exclusively with non-periodic Molecules, I'm not sure I have a complete understanding of the differences that would be involved with Structures.

So, I think that there are a couple of different levels of complexity here, and I'm not sure there's a single solution that elegantly addresses all of them.

- For simple functional group substitution, I agree with @mkhorton that using the StructureGraph/MoleculeGraph classes is a bit overkill. I haven't read the SiteCollection class that closely, but I think it should be possible to extend & generalize Molecule.substitute().
- For swapping functional groups, as @shyuep mentioned, we could also stay at the level of Molecule.substitute() for simple functional groups. But, if you want to do more complicated substitutions - for instance, changing a functional group that is attached to the larger molecule in more than one place - then the graph representation simplifies things significantly. I already have a somewhat more robust molecule-splitting function in MoleculeGraph that could allow for this (MoleculeGraph.split_molecule_subgraphs()).
- For my workflows, I've been simultaneously substituting functional groups into reactant and product molecules. This use case requires the graph representation (which reduces the problem to a search for subgraph isomorphisms) to be at all reasonable. But, I don't know if that use case of mapping substitutions to multiple related Structures would be needed.

In summary, it's probably best to have functional group substitution both at the level of Molecule/Structure (bundled into SiteCollection or otherwise) and MoleculeGraph/StructureGraph. This wasn't at all what the original Issue was about, but if we're going to be adding even more similar functionality between the two Graph functions than what already exists, we might also want to refactor (again) and make some sort of base class for the two.
